Hydra/Serializer/PaginatedCollectionEnhancer.php

```diff
@@ -0,0 +1,166 @@
+<?php
+
+/*
+ * This file is part of the DunglasApiBundle package.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Dunglas\ApiBundle\Hydra\Serializer;
+
+use Dunglas\ApiBundle\Api\ResourceInterface;
+use Dunglas\ApiBundle\Api\ResourceResolver;
+use Dunglas\ApiBundle\Model\PaginatorInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+use Symfony\Component\Serializer\Normalizer\SerializerAwareNormalizer;
+use Symfony\Component\Serializer\SerializerInterface;
+
+/**
+ * Enhance the result of collection by enabling pagination.
+ *
+ * @author Samuel ROZE <samuel.roze@gmail.com>
+ */
+class PaginatedCollectionEnhancer extends SerializerAwareNormalizer implements NormalizerInterface
+{
+    /**
+     * @var string
+     */
+    const HYDRA_PAGED_COLLECTION = 'hydra:PagedCollection';
+
+    /**
+     * @var CollectionNormalizer
+     */
+    private $collectionNormalizer;
+    /**
+     * @var ResourceResolver
+     */
+    private $resourceResolver;
+
+    /**
+     * @param NormalizerInterface $collectionNormalizer
+     * @param ResourceResolver    $resourceResolver
+     */
+    public function __construct(NormalizerInterface $collectionNormalizer, ResourceResolver $resourceResolver)
+    {
+        $this->collectionNormalizer = $collectionNormalizer;
+        $this->resourceResolver = $resourceResolver;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function normalize($object, $format = null, array $context = array())
+    {
+        $data = $this->collectionNormalizer->normalize($object, $format, $context);
+        if (isset($context['json_ld_sub_level']) || !$object instanceof PaginatorInterface) {
+            return $data;
+        }
+
+        $resource = $this->resourceResolver->guessResource($object, $context);
+        list($parts, $parameters) = $this->parseRequestUri($resource, $context['request_uri']);
+
+        $data['@type'] = self::HYDRA_PAGED_COLLECTION;
+
+        $currentPage = $object->getCurrentPage();
+        $lastPage = $object->getLastPage();
+
+        if (1. !== $currentPage) {
+            $previousPage = $currentPage - 1.;
+            $data['hydra:previousPage'] = $this->getPageUrl($resource, $parts, $parameters, $previousPage);
+        }
+
+        if ($currentPage !== $lastPage) {
+            $data['hydra:nextPage'] = $this->getPageUrl($resource, $parts, $parameters, $currentPage + 1.);
+        }
+
+        $data['hydra:totalItems'] = $object->getTotalItems();
+        $data['hydra:itemsPerPage'] = $object->getItemsPerPage();
+        $data['hydra:firstPage'] = $this->getPageUrl($resource, $parts, $parameters, 1.);
+        $data['hydra:lastPage'] = $this->getPageUrl($resource, $parts, $parameters, $lastPage);
+
+        // Reorder the `hydra:member` key to the end
```

Comment:
I don't think it's necessary because I don't think that the JSON fields ordering are important. BTW, right know, based on how the bundle is tested, it's massively based on this order.